회원가입 기능 개발

Habist opened the pull request:
## 작업 내용
- 요청 정보의 유효성 검증을 위해 validation 추가
- 사용자 회원가입 요청을 받는 컨트롤러 추가
- 요청에 따른 비즈니스 로직 호출을 위한 서비스 추가
- DB 조작을 위한 Mapper 추가
- 회원가입 정보를 담는 User 모델 추가
- 비밀번호 암호화를 위한 SHA-256 Utility 추가
- 회원 정보를 저장하는 User Table 생성 쿼리 추가
- h2 DB 사용을 위한 옵션 추가


>  참조이슈
> #5 [기능] 회원가입 기능

Habist committed:
chore: [#5] maven 의존성 추가

2.3 버전 이상의 spring boot web starter에 validation starter를
더 이상 포함하지 않아 추가하였습니다.

Habist committed:
feat: [#5] 사용자 회원가입 기능 개발

사용자 회원가입 요청을 받는 컨트롤러를 추가했습니다.
요청에 따른 비즈니스 로직 호출을 위해 서비스를 추가했습니다.
DB 조작을 위한 Mapper를 추가했습니다.
회원가입 정보를 담는 User 모델을 추가했습니다.
비밀번호 암호화를 위한 SHA-256 Utility를 추가했습니다.
회원 정보를 저장하는 User Table 생성 쿼리를 추가했습니다.
h2 DB 사용을 위해 옵션을 추가했습니다.

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final UserService userService;
+
+    public UserController(UserService userService) {
```
이런 코드는 `lombok`을 사용하면 간소화시킬 수 있습니다~ `lombok`이 무엇인지 알아보시면 좋겠습니다

Habist replied:
확인해보니 lombok에는 final 키워드 또는 NonNull 애노테이션이 선언된 필드를 매개변수로 받는 생성자를 RequiredArgsConstructor 애노테이션을 통해서 간소화시킬 수 있네요! 
리뷰 반영했습니다!

f-lab-dev commented on src/main/java/com/campool/model/User.java:
```diff
@@ -0,0 +1,37 @@
+@Getter
+@Setter
+@ToString
+public class User {
```
유저에 대한 정보를 주고받을 때 이 클래스 하나로 공통적으로 사용하시려는 것 같은데 아래와 같은 문제가 발생할 수 있습니다.

- 굳이 넘겨주지 않아도 되는 정보를 실수로 넘기게 될 가능성이 있습니다
- 공통적으로 쓸거니까 setter를 선언함으로써 immutable이 깨지게 됩니다
- 요청, 응답용으로 사용할 때 `어떤 값을 넘기고 어떤 값이 리턴된다`라는 명세가 명확하게 보이지 않게 됩니다

Habist replied:
그러면 요청과 넘겨줘야하는 정보에 따라 각각의 클래스를 만들어 사용하는 것이 괜찮나요?

f-lab-dev replied:
네 다만 그렇게 하면 클래스의 개수가 엄청 많아지겠죠~ 그것도 어떻게 해결하는게 좋을지 고민해보시면 좋을 것 같습니다

f-lab-dev commented on src/main/java/com/campool/model/User.java:
```diff
@@ -0,0 +1,37 @@
+    private String email;
+
+    @NotBlank
+    @Pattern(regexp = "[0-9]{10,11}")
```
만약 잘못된 값이 들어왔을 때 에러 메세지를 보내줄 수 있는 옵션이 있습니다. 그것도 고려해보시면 좋겠네요~

Habist replied:
입력 시 규칙에 대한 명확한 인지를 위해 에러 메시지를 추가해주는 것이 좋겠네요!

리뷰 반영하겠습니다~

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public void add(User user) {
+        if (userMapper.findById(user.getId()) == null) {
```
이 부분은 메소드로 분리하면 아이디 중복확인 기능의 재사용성이 높아지지 않을까요?

Habist replied:
넵 아이디 중복확인 기능은 다른 곳에서도 사용될 여지가 높아 재사용을 위해 메소드로 분리하는 것이 좋을 것 같네요~

리뷰 반영했습니다!

f-lab-dev commented on src/main/java/com/campool/utils/SHA256Utility.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.security.MessageDigest;
+
+public class SHA256Utility {
```
이 클래스는 암호화 방식이 바뀌게 된다면 그런 변화에 유연하게 대처할 수 있을까요?

Habist replied:
변경되는 상황에서는 유연하게 대처가 힘들 것 같네요.
암호화를 사용하는 부분을 인터페이스에 의존시키고 암호화 방식에 따라 다르게 구체화한 클래스를 주입해주는 방식으로 결합도를 낮추는 게 좋을까요?

f-lab-dev replied:
네 한번 그렇게 개선해보시면 좋겠네요

f-lab-dev commented on src/main/java/com/campool/utils/SHA256Utility.java:
```diff
@@ -0,0 +1,29 @@
+            }
+            return sb.toString();
+        } catch (Exception e) {
+            return null;
```
여기선 예외상황에 null을 리턴하고 있습니다. 일단 값을 리턴한다는 것은 정상적인 코드 흐름으로 처리하겠다는건데 실수로 null 값을 DB에 넣어버리는 등의 버그가 발생할 가능성이 생기지 않을까요?

Habist replied:
USER 테이블 password 컬럼의 제약 조건으로 Not Null을 사용해 null에 대한 처리를 DB 수준에서 처리하도록 만들려고 생각했었습니다. 하지만 애초에 암호화 처리 과정에서의 예외 상황이니 DB가 아닌 암호화를 사용하는 단계에서 명확한 예외 인지가 필요하다고 판단이 들어 예외를 던지도록 수정했습니다!

Habist committed:
refactor: [#5 - 리뷰반영] constructor injection 간소화

생성자를 통한 의존성 주입 부분을 Lombok의 RequiredArgsConstructor 애노테이션을 통해
간소화하였습니다.

Habist committed:
refactor: [#5 - 리뷰반영] 사용자 아이디 중복확인 코드 분리

중복확인 코드의 재사용을 위해 메소드로 분리하였습니다.

Habist committed:
refactor: [#5 - 리뷰반영] 표현식과 다른 입력값에 대한 에러 메시지 추가

User 클래스의 telephone 필드 표현식과 다른 입력값에 대한 에러 메시지를 추가했습니다.

Habist committed:
improve: [#5 - 리뷰반영] 암호화 처리 중 예외 발생 시 런타임 예외 생성

예외 상황 가시성 위해 암호화 처리 중 예외가 발생하면
null 반환이 아닌 런타임 예외를 던지도록 변경했습니다.

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -3,18 +3,18 @@
 @RestController
 public class UserController {
 
+    @NonNull
```
Constructor injection을 사용한다면 이 어노테이션을 달지 않더라도 애초에 null이 안들어가지 않을까요?

Habist replied:
NonNull 애노테이션을 달지 않으면 null이 변수에 할당되는 상황에 대해 검증하지 않아서 null이 들어갈 가능성이 생깁니다.
스프링이 injection을 해줄때는 의존성에 대한 null 체크를 해주지만 해당 클래스의 인스턴스를 직접 생성하는 경우에는 null 할당 방지를 보장할 수 없기 때문에 null 할당 가능성을 최대한 제어하는 용도로 NonNull을 사용했습니다!

f-lab-dev replied:
그런 의도셨군요~ 👍

f-lab-dev commented on src/main/java/com/campool/service/UserService.java:
```diff
@@ -15,12 +15,16 @@
     public void add(User user) {
-        if (userMapper.findById(user.getId()) == null) {
+        if (isDuplicate(user.getId())) {
+            throw new RuntimeException("중복된 아이디가 존재합니다.");
```
`RuntimeException` 보다 구체적인 타입의 예외를 던져주면 더 좋을 것 같습니다~

Habist replied:
중복의 의미를 나타낼 수 있는 예외로 변경하였습니다!

f-lab-dev commented on src/main/java/com/campool/utils/SHA256Utility.java:
```diff
@@ -22,7 +22,7 @@ public static String encrypt(String msg) {
             return sb.toString();
         } catch (Exception e) {
-            return null;
+            throw new RuntimeException(e);
```
여기 또한 예외 타입과 메세지를 구체적으로 던져주는게 좋을 것 같네요

Habist replied:
이 부분은 구체적인 예외 타입을 catch하도록 하고 checked exception을 unchecked exception으로 예외 포장하는 방법으로 변경하였습니다.

f-lab-dev commented on src/main/java/com/campool/model/User.java:
```diff
@@ -0,0 +1,37 @@
+    private String email;
+
+    @NotBlank
+    @Pattern(regexp = "[0-9]{10,11}", message = "- 을 제외한 숫자 10자리 또는 11자리를 입력해주세요.")
```
여기뿐 아니라 다른 곳에도 메세지를 넣어줘야하지 않을까요? 리뷰를 받을때는 리뷰를 받은곳 뿐만 아니라 비슷한 곳들이 있는지 다른 소스코드들도 다 찾아보면서 고쳐주세요~

Habist replied:
넵 다른 비슷한 부분들도 확인하고 수정하겠습니다~

Habist committed:
improve: [#5] 리뷰 반영

암호화 모듈 DIP 적용
회원가입에 필요한 모델로 변경
회원가입 모델 유효성 검증 시 에러 메시지 추가
회원가입 모델에 비밀번호 암호화 로직 추가
구체적 에러 타입과 메시지로 수정

f-lab-dev commented on src/main/java/com/campool/encrypt/SHA256Encryptor.java:
```diff
@@ -1,16 +1,14 @@
+import org.springframework.stereotype.Component;
 
-public class SHA256Utility {
+@Component
```
`@Component`와 `@Service`의 차이점이 무엇일까요?

Habist replied:
가장 큰 차이점은 선언된 클래스가 사용되는 layer에 대한 **명확성**이 확보된다는 점입니다.
그래서 비즈니스 서비스에 대한 처리를 하는 클래스를 명확하게 `@Service`를 통해 구분할 수 있습니다.
또한 AOP 관점에서는 PointCut에 적용할 구체적인 타깃으로 사용될 수도 있습니다.

저는 암호화를 담당하는 클래스는 어느 계층에도 속해있지 않고, 그저 빈으로 등록되어 다양한 곳에서 사용될 수 있다고
판단하여 `@Component`를 선언했습니다!

f-lab-dev reviewed (approved):
질문드린 부분만 답변주시고 머지하셔도 됩니다~
